### ISSUE TITLE
Offer backward compatible Identity rule mechanism

### ISSUE THREAD

**Issue by BenjaminAbt** [dotnet/aspnetcore#39194]
It's unclear to me, unfortunately, if this is a design flaw (in which case it would be a feature request to fix that) or if there is a built-in solution (in which case this is a question).

### Is your feature request related to a problem? Please describe.

ASP.NET Core Identity has the possibility to define rules for e.g. usernames, which is usually widely used.
However, it is also often the case in the real world that these rules can change - for example, special characters are subsequently banned in the username.

But, if the rules are changed, many identity operations can end with exceptions on existing users.
For example, a **password** can no longer be changed if a **username** does not match the new rules....(!).

> Password change exception: Identity Error InvalidUserName: Username 'User Name' is invalid, can only contain letters or digits.

It is forced to update all usernames when the rules are changed. Why?

Unfortunately, this only becomes apparent when the corresponding updates have been rolled out. And then a manual intervention of the developers in the production system is necessary, e.g. to be able to fight password leaks (current case).
The user himself is blocked.

### Describe the solution you'd like
A backward compatible solution would be better or a way to hook into the validation process for such cases.

**Comment by blowdart** [comment]
I don't think we ever considered this; we assume that if you add rules for validity they must be enforced everywhere,.

It would be a breaking change, and an interesting one at that. Rather than break what may work for others, would a setting work for you here? For example, some setting on identity that says, "Ignore username validity checks when updating a user"? It would be off by default, but you could configure it on startup. That way we don't change the current behaviour, but you can opt-in.

**Comment by BenjaminAbt** [comment]
I guess I would prefer a validation interface that would be implemented similarly to the rest of the Identity providers.
That means a general interface and a default implementation that can be replaced (like the UserManager).

`public class CustomUserManager : UserManager<MyUser>`

but for the validation

```cs
public class CustomUserValidator : UserValidator<MyUser>
{
   public CustomUserValidator(IOptions<IdentityOptions> optionsAccessor, ....) : base(optionsAccessor) { }
 
   public IdentityValidationResult ValidateUsername(MyUser user, string userName)
   {
      ...
   }

   public IdentityValidationResult ValidatePassword(MyUser user, string newPassword)
   {
      ...
   }
}
```

The mechanism would therefore be very similar to the current ASP.NET Identity structure, the default behavior wont change - but there would be maximum flexibility for customization.
A general option could of course be introduced in addition.

**Comment by blowdart** [comment]
@BenjaminAbt That seems complicated to me, how would you use that? You wouldn't be able to tell what the operation was, so you couldn't do, for example, do not validate the username when changing the password only, unless you're expecting us to just call validatepassword, which, of course, already exists kind of, with the validation rules.

We prefer a flag option in identity settings, it's easier to understand.

**Comment by HaoK** [comment]
@BenjaminAbt you should be able to do this already via a custom user manager that overrides ChangePasswordAsync to call your own version of UpdateUserWithOutValidation

**Comment by BenjaminAbt** [comment]
> how would you use that? 

Yes, it probably couldn't be done with the current system without major changes.
Currently, the entire user is always validated (`ValidateUserAsync`) without knowing what has actually changed.

@HaoK thanks for the hint.
Currently, we are bypassing the entire user management process. We actually only use the identity system to have a secure login behavior, 2FA for free and not have to do certain tasks like password hashing ourselves.

Everything else is disabled because it's not applicable (for us / my customers).
